cpp/unit/action.cpp

```diff
@@ -79,6 +96,130 @@ void UnitAction::move_to(Unit &target, bool use_range) {
 	this->entity->invoke(cmd);
 }
 
+TargetAction::TargetAction(Unit *u, graphic_type gt, UnitReference r, coord::phys_t rad)
+	:
+	UnitAction(u, gt),
+	target{r},
+	repath_attempts{10},
+	end_action{false},
+	radius{rad} {
+	this->update_distance();
+}
+
+TargetAction::TargetAction(Unit *u, graphic_type gt, UnitReference r)
+	:
+	TargetAction(u, gt, r, adjacent_range(u)) {
+}
+
+void TargetAction::update(unsigned int time) {
+	auto target_ptr = update_distance();
+	if (!target_ptr) {
+		return;
+	}
+
+	if (this->entity->has_attribute(attr_type::gatherer)) {
+
+		// the gatherer attributes attached to the unit
+		// are used to modify the graphic
+		auto &gatherer_attr = this->entity->get_attribute<attr_type::gatherer>();
+
+		if (target_ptr->has_attribute(attr_type::resource)) {
+
+			// target contains resources
+			// set gatherer graphics
+			auto &resource_attr = target_ptr->get_attribute<attr_type::resource>();
+			gatherer_attr.current_type = resource_attr.resource_type;
+
+			// check graphics are available
+			auto class_type = target_ptr->unit_class;
+			if (gatherer_attr.graphics.count(class_type) > 0) {
+				this->entity->graphics = &gatherer_attr.graphics[class_type]->graphics;
+			}
+		}
+		else if (this->name() == "build" &&
+		         target_ptr->has_attribute(attr_type::building)) {
+
+			// set builder graphics if available
+			if (gatherer_attr.graphics.count(gamedata::unit_classes::BUILDING) > 0) {
+				this->entity->graphics = &gatherer_attr.graphics[gamedata::unit_classes::BUILDING]->graphics;
+			}
+		}
+		else if (this->name() == "attack" &&
+		         this->entity->has_attribute(attr_type::attack)) {
+
+			// target does not have resource
+			auto &attack_attr = this->entity->get_attribute<attr_type::attack>();
+			this->entity->graphics = attack_attr.attack_graphic_set;
+		}
+	}
+
+	// set direction unit should face
+	this->face_towards(target_ptr->location->pos.draw);
+
+	// move to within the set radius
+	if (this->dist_to_target <= this->radius) {
+
+		// the derived class controls what to
+		// do when in range of the target
+		this->update_in_range(time, target_ptr);
+		this->repath_attempts = 10;
+	}
+	else if (this->repath_attempts) {
+
+		// out of range so try move towards
+		// if this unit has a move ability
+		this->move_to(*target_ptr);
+		this->repath_attempts -= 1;
+	}
+	else {
+
+		// unit is stuck
+		this->end_action = true;
+	}
+}
+
+void TargetAction::on_completion() {}
```

Comment:
its is just a todo for now,
gathers go to idle when its resource runs out, they will have a find new resource function in the n_completion in future